### PR TITLE
Neural Graph Actions

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**
This PR introduces the following graph methods:

 - [x] to(device) - moves all modules in a graph to a given device
 - [x] data_loader_configure(...) - enables parametrization of PyTorch's DataLoader associated with a given graph/DataLayer/Dataset
 - [x] get_batch() - returns a batch (as tuple or dictionary)
 - [x] forward(batch) - forward pass through the graph 
 - [x] backward(losses=[]) - backward pass, if losses not provided - backpropagates from all outputs of all losses in a given graph
 - [x] parameters() - yields parameters of graph modules and all their submodules
 - [x] named_parameters() - yields (names: parameters) of graph modules and all their submodules

New graph properties:
 - [x] is_trainable - checks if graph is "trainable", which means that a) the graph is in training mode and b) has at least one Loss Neural Modules.

As a result one can use graphs directly in training or inference.

Exemplary training loop::

```
        # Iterate over the whole dataset - in batches.
        for step, batch in enumerate(training_graph.get_batch()):
            # Reset the gradients.
            opt.zero_grad()
            # Forward pass.
            outputs = training_graph.forward(batch)
            # Print loss.
            if step % freq == 0:
                logging.info("Epoch: {} Step: {} Training Loss: {}".format(epoch, step, outputs.loss))
            # Backpropagate the gradients.
            training_graph.backward()
            # Update the parameters.
            opt.step()
```

Two refactored examples:
 - [x] NLP: rule_based_policy_managed - pure inference using graph.forward() in evaluation mode, on CPU
 - [x] CV: mnist_lenet5_image_classification -  training loop using graph forward + backward, on GPU, evaluation using second graph in evaluation mode on GPU

The LeNet-5 training was tested/works on both on a single GPU and two GPUs using DDP:
Screen Shot 2020-06-14 at 2 52 51 PM

Fixes:
 - [x] CV: removed `to(device)` calls from all modules

### CODE REVIEW

**Comment by tkornuta-nvidia** [comment]
ok, will do it as part of the next PR which will use those functions in the notebook.
